src/topology/subset_properties.lean

```diff
@@ -89,61 +89,118 @@ lemma compact_iff_ultrafilter_le_nhds {s : set α} :
-    ⟨t, ht₁, (ht₂ : a ∈ t)⟩ := hc₂ ha
+    ⟨_, ⟨i, rfl⟩, (ha : a ∈ U i)⟩ := hsU ha
   in
-  have f ≤ principal (-t),
-    from infi_le_of_le ⟨{t}, by rwa singleton_subset_iff, finite_insert _ finite_empty⟩ $
-      principal_mono.mpr $
-        show s - ⋃₀{t} ⊆ - t, begin rw sUnion_singleton; exact assume x ⟨_, hnt⟩, hnt end,
-  have is_closed (- t), from is_open_compl_iff.mp $ by rw lattice.neg_neg; exact hc₁ t ht₁,
-  have a ∈ - t, from is_closed_iff_nhds.mp this _ $ ne_bot_of_le_ne_bot h $
-    le_inf inf_le_right (inf_le_left_of_le ‹f ≤ principal (- t)›),
-  this ‹a ∈ t›
+  have f ≤ principal (- U i),
+    from infi_le_of_le {i} $ principal_mono.mpr $ show s - _ ⊆ - U i, by simp [diff_subset_iff],
+  have is_closed (- U i), from is_open_compl_iff.mp $ by rw lattice.neg_neg; exact hUo i,
+  have a ∈ - U i, from is_closed_iff_nhds.mp this _ $ ne_bot_of_le_ne_bot h $
+    le_inf inf_le_right (inf_le_left_of_le ‹f ≤ principal (- U i)›),
+  this ‹a ∈ U i›
+
+/-- For every family of closed sets whose intersection avoids a compact set,
+there exists a finite subfamily whose intersection avoids this compact set. -/
+lemma compact.elim_finite_subfamily_closed {s : set α} {ι : Type v} (hs : compact s)
+  (Z : ι → set α) (hZc : ∀i, is_closed (Z i)) (hsZ : s ∩ (⋂ i, Z i) = ∅) :
+  ∃ t : finset ι, s ∩ (⋂ i ∈ t, Z i) = ∅ :=
+let ⟨t, ht⟩ := hs.elim_finite_subcover (λ i, - Z i) hZc
+  (by simpa only [subset_def, not_forall, eq_empty_iff_forall_not_mem, set.mem_Union,
+    exists_prop, set.mem_inter_eq, not_and, iff_self, set.mem_Inter, set.mem_compl_eq] using hsZ)
+    in
+⟨t, by simpa only [subset_def, not_forall, eq_empty_iff_forall_not_mem, set.mem_Union,
+    exists_prop, set.mem_inter_eq, not_and, iff_self, set.mem_Inter, set.mem_compl_eq] using ht⟩
+
+/-- Cantor's intersection theorem:
+if the union of a directed family of nonempty closed sets `Z i` is compact,
+then the intersection of the `Z i` is nonempty. -/
+lemma compact.nonempty_Inter_of_compact_Union_of_directed_nonempty_closed
+  {ι : Type v} [hι : nonempty ι] (Z : ι → set α) (hZ : compact (⋃ i, Z i))
+  (hZd : directed (⊇) Z) (hZn : ∀ i, (Z i).nonempty) (hZc : ∀ i, is_closed (Z i)) :
+  (⋂ i, Z i).nonempty :=
+begin
+  rw ← ne_empty_iff_nonempty,
+  intro H,
+  obtain ⟨t, ht⟩ : ∃ (t : finset ι), ((⋃ i, Z i) ∩ ⋂ (i ∈ t), Z i) = ∅,
+    from hZ.elim_finite_subfamily_closed Z hZc (by rw [H, inter_empty]),
+  obtain ⟨i₀, hi₀⟩ : ∃ i₀ : ι, ∀ i ∈ t, Z i ⊇ Z i₀,
+    from directed.finset_le (by apply_instance) hZd t,
+  suffices : ((⋃ i, Z i) ∩ ⋂ (i ∈ t), Z i).nonempty,
+  { rw ← ne_empty_iff_nonempty at this, contradiction },
+  refine nonempty.mono _ (hZn i₀),
+  exact subset_inter (subset_Union Z i₀) (subset_bInter hi₀)
+end
 
+/-- Cantor's intersection theorem for sequences indexed by `ℕ`:
+if the union of a decreasing sequence of nonempty closed sets `Z i` is compact,
+then the intersection of the `Z i` is nonempty. -/
+lemma compact.nonempty_Inter_of_compact_Union_of_sequence_nonempty_closed
```

Comment:
Here it is again not obvious what the right formulation is: requiring that `Z 0` is compact and that all the `Z i` are closed, as you do, or directly requiring that all of them are `compact`? (The two formulations are equivalent). Maybe give both versions?

Reply:
In general compact doesn't imply closed. And in the setting where it does, I guess it is a 1-liner to get the right hypothesis. Don't you think so?

Reply:
Fair enough.